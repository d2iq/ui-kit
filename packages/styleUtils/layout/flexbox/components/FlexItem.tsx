import * as React from "react";
import { css } from "@emotion/css";
import {
  atMediaUp,
  BreakpointConfig
} from "../../../../shared/styles/breakpoints";
import { getResponsiveStyle } from "../../../../shared/styles/styleUtils";
import { flexItem } from "../../../../shared/styles/styleUtils/layout/flexbox";

type FlexStrategy = "shrink" | "grow";

interface FlexItemProps {
  className?: string;
  /**
   * Whether the flex item should grow or shrinkwrap to it's children. Can be set for all viewport sizes, or configured to have different values at different viewport width breakpoints
   */
  flex?: BreakpointConfig<FlexStrategy>;
  /**
   * The rate at which a flex item should grow. Can be set for all viewport sizes, or configured to have different values at different viewport width breakpoints
   */
  growFactor?: BreakpointConfig<React.CSSProperties["flexGrow"]>;
  /**
   * The position at which a flex item should be visually rendered. Can be set for all viewport sizes, or configured to have different values at different viewport width breakpoints
   */
  order?: BreakpointConfig<React.CSSProperties["order"]>;
  /**
   * Which HTML tag to render the component with
   */
  tag?: keyof React.ReactHTML;
}

const getResponsiveFlexItemStyles = flexVal =>
  typeof flexVal !== "object"
    ? flexItem(flexVal)
    : Object.keys(flexVal).map(breakpoint =>
        atMediaUp[breakpoint](flexItem(flexVal[breakpoint]))
      );

const FlexItem: React.FC<FlexItemProps> = ({
  children,
  className,
  flex,
  growFactor,
  order,
  tag = "div"
}) => {
  const FlexItemEl = tag;

  return (
    <FlexItemEl
      className={css`
        ${getResponsiveFlexItemStyles(flex)};
        ${getResponsiveStyle("flex-grow", growFactor)};
        ${getResponsiveStyle("order", order)};
        ${className};
      `}
      data-cy="flexItem"
    >
      {children}
    </FlexItemEl>
  );
};

FlexItem.defaultProps = {
  flex: "grow"
};

export default FlexItem;
