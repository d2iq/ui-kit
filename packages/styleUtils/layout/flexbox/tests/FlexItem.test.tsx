import React from "react";
import { shallow } from "enzyme";
import { createSerializer } from "@emotion/jest";
import toJson from "enzyme-to-json";

import FlexItem from "../components/FlexItem";

expect.addSnapshotSerializer(createSerializer());

describe("FlexItem", () => {
  it("renders default", () => {
    const component = shallow(<FlexItem>content</FlexItem>);

    expect(toJson(component)).toMatchSnapshot();
  });
  it("renders with all props", () => {
    const component = shallow(
      <FlexItem flex="grow" growFactor={2} order={2}>
        content
      </FlexItem>
    );

    expect(toJson(component)).toMatchSnapshot();
  });
  it("renders with responsive props", () => {
    const component = shallow(
      <FlexItem
        flex={{ default: "shrink", medium: "grow" }}
        growFactor={{ default: 1, medium: 2 }}
        order={{ default: 1, medium: 2 }}
      >
        content
      </FlexItem>
    );

    expect(toJson(component)).toMatchSnapshot();
  });
});
