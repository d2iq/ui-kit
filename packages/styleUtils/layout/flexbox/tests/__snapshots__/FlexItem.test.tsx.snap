// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlexItem renders default 1`] = `
.emotion-0 {
  box-sizing: border-box;
  -webkit-flex-basis: 0;
  -ms-flex-preferred-size: 0;
  flex-basis: 0;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  min-width: 0;
  width: auto;
}

<div
  className="emotion-0"
  data-cy="flexItem"
>
  content
</div>
`;

exports[`FlexItem renders with all props 1`] = `
.emotion-0 {
  box-sizing: border-box;
  -webkit-flex-basis: 0;
  -ms-flex-preferred-size: 0;
  flex-basis: 0;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  min-width: 0;
  width: auto;
  -webkit-box-flex: 2;
  -webkit-flex-grow: 2;
  -ms-flex-positive: 2;
  flex-grow: 2;
  -webkit-order: 2;
  -ms-flex-order: 2;
  order: 2;
}

<div
  className="emotion-0"
  data-cy="flexItem"
>
  content
</div>
`;

exports[`FlexItem renders with responsive props 1`] = `
@media (min-width: 0) {
  .emotion-0 {
    box-sizing: border-box;
    -webkit-flex-basis: auto;
    -ms-flex-preferred-size: auto;
    flex-basis: auto;
    -webkit-box-flex: 0;
    -webkit-flex-grow: 0;
    -ms-flex-positive: 0;
    flex-grow: 0;
    -webkit-flex-shrink: 0;
    -ms-flex-negative: 0;
    flex-shrink: 0;
    width: initial;
  }
}

@media (min-width: 900px) {
  .emotion-0 {
    box-sizing: border-box;
    -webkit-flex-basis: 0;
    -ms-flex-preferred-size: 0;
    flex-basis: 0;
    -webkit-box-flex: 1;
    -webkit-flex-grow: 1;
    -ms-flex-positive: 1;
    flex-grow: 1;
    min-width: 0;
    width: auto;
  }
}

@media (min-width: 0) {
  .emotion-0 {
    -webkit-box-flex: 1;
    -webkit-flex-grow: 1;
    -ms-flex-positive: 1;
    flex-grow: 1;
  }
}

@media (min-width: 900px) {
  .emotion-0 {
    -webkit-box-flex: 2;
    -webkit-flex-grow: 2;
    -ms-flex-positive: 2;
    flex-grow: 2;
  }
}

@media (min-width: 0) {
  .emotion-0 {
    -webkit-order: 1;
    -ms-flex-order: 1;
    order: 1;
  }
}

@media (min-width: 900px) {
  .emotion-0 {
    -webkit-order: 2;
    -ms-flex-order: 2;
    order: 2;
  }
}

<div
  className="emotion-0"
  data-cy="flexItem"
>
  content
</div>
`;
